src/core/qgsvectorlayercache.cpp

```diff
@@ -231,6 +237,7 @@ void QgsVectorLayerCache::attributeAdded( int field )
 {
   Q_UNUSED( field )
   mCachedAttributes.append( field );
+  mFullCache = false;
```

Comment:
Actually I've reintroduced this fix, because it turns out it's valid. If the cache is invalidated for any reason (eg by adding an attribute to the table) and mFullCache is true, then there's no choice but to rebuild the entire cache from the table. This would be a performance killer, so my newly introduced fix makes it clear in the documentation that it's necessary to listen out for the invalidated signal and if required then handle rebuilding the full cache in a nice responsive way (eg with progress feedback). I've updated the attribute table dialog to suit.